docs/architecture/003-handler-implementation.md

```diff
@@ -0,0 +1,552 @@
+# ADR 003: Handler implementation
```

Comment:
I propose we use the term "protocol" to replace "handlers".
The "message processing logic" for a given codebase is effectively implementing the "protocol".

The term "protocol" is still imperfect, but easier to talk about than "message processing logic" and less contentious than "handler". A protocol consists of processing functions `process_msg_x`, `process_msg_y` etc. so that any input message is processed by such a function.